lib/ClangImporter/ClangImporter.cpp

```diff
@@ -5828,3 +5828,295 @@ void ClangImporter::diagnoseMemberValue(const DeclName &name,
+      return method->isCopyAssignmentOperator();
+  });
+
+  assert(copyAssignOp != decl->decls_end());
+  return cast<clang::CXXMethodDecl>(*copyAssignOp)->isDefaulted();
+}
+
+/// Recursively checks that there are no user-provided copy constructors or
+/// destructors in any fields or base classes.
+/// Does not check C++ records with specific API annotations.
+static bool isSufficientlyTrivial(const clang::CXXRecordDecl *decl) {
+  // Probably a class template that has not yet been specialized:
+  if (!decl->getDefinition())
+    return true;
+
+  if ((decl->hasUserDeclaredCopyConstructor() &&
+       !copyConstructorIsDefaulted(decl)) ||
+      (decl->hasUserDeclaredCopyAssignment() &&
+       !copyAssignOperatorIsDefaulted(decl)) ||
+      (decl->hasUserDeclaredDestructor() &&
+       !decl->getDestructor()->isDefaulted()))
+    return false;
+
+  auto checkType = [](clang::QualType t) {
+    if (auto recordType = dyn_cast<clang::RecordType>(t.getCanonicalType())) {
+      if (auto cxxRecord =
+              dyn_cast<clang::CXXRecordDecl>(recordType->getDecl())) {
+        if (hasImportAsRefAttr(cxxRecord) || hasOwnedValueAttr(cxxRecord) ||
+            hasUnsafeAPIAttr(cxxRecord))
+          return true;
+
+        if (!isSufficientlyTrivial(cxxRecord))
+          return false;
+      }
+    }
+
+    return true;
+  };
+
+  for (auto field : decl->fields()) {
+    if (!checkType(field->getType()))
+      return false;
+  }
+
+  for (auto base : decl->bases()) {
+    if (!checkType(base.getType()))
+      return false;
+  }
+
+  return true;
+}
+
+/// Checks if a record provides the required value type lifetime operations
+/// (copy and destroy).
+static bool hasRequiredValueTypeOperations(const clang::CXXRecordDecl *decl) {
+  if (auto dtor = decl->getDestructor()) {
+    if (dtor->isDeleted() || dtor->getAccess() != clang::AS_public) {
+      return false;
+    }
+  }
+
+  // If we have no way of copying the type we can't import the class
+  // at all because we cannot express the correct semantics as a swift
+  // struct.
+  if (llvm::any_of(decl->ctors(), [](clang::CXXConstructorDecl *ctor) {
+        return ctor->isCopyConstructor() &&
+               (ctor->isDeleted() || ctor->getAccess() != clang::AS_public);
+      }))
+    return false;
+
+  return true;
+}
+
+CxxRecordSemanticsKind
+CxxRecordSemantics::evaluate(Evaluator &evaluator,
+                             CxxRecordSemanticsDescriptor desc) const {
+  const auto *decl = desc.decl;
+
+  if (hasImportAsRefAttr(decl)) {
+    return CxxRecordSemanticsKind::Reference;
+  }
+
+  // TODO: diagnose if the decl also has any attrs.
+  if (!hasRequiredValueTypeOperations(decl)) {
+    return CxxRecordSemanticsKind::UnsafeLifetimeOperation;
+  }
+
+  if (hasUnsafeAPIAttr(decl)) {
+    return CxxRecordSemanticsKind::ExplicitlyUnsafe;
+  }
+
+  if (hasOwnedValueAttr(decl)) {
+    return CxxRecordSemanticsKind::Owned;
+  }
+
+  if (hasIteratorAPIAttr(decl)) {
```

Comment:
I don't think we should require the user to annotate all iterators with an attribute. Swift can detect that the type is an iterator statically without an attribute by doing `clangRecordDecl->lookup(getClangASTContext().Idents.get("iterator_category"))` – if there is a subtype called `iterator_category`, then this type is an iterator (see `swift::conformToCxxIteratorIfNeeded` from https://github.com/apple/swift/pull/60001).

If the iterator type doesn't define `iterator_category`, then the user can add an attribute, but I don't think we should require attributes for something we can correctly deduce statically.

Reply:
As discussed offline: we will use both to determine what's an iterator.